### ISSUE TITLE
ThreadLocal memory leak

### ISSUE THREAD

**Issue by amichair** [OpenHFT/Chronicle-Queue#305]
SingleChronicleQueue uses a couple of ThreadLocals which are never removed. This causes a whole tree of references to be leaked permanently after the queue is closed.

Additionally, if there is a storeFileListener set, there is no way to remove it either, so combined with this bug it can keep an entire application stuck in memory (in our case in an OSGi enviroment, where a bundle/application can be reloaded, the old bundle/application lingers on indefinitely).

Please make sure to remove all ThreadLocals when they are no longer needed, in this class and anywhere else they are used. You may also consider adding a way to remove a StoreFileListener, or doing so internally in appropriate places if it's still needed after the above fix.

**Comment by peter-lawrey** [comment]
The ThreadLocal are local to the queue. Are these being retained after the
Queue is being GCed?

**Comment by amichair** [comment]
I think the queue will not be GCed, since the threads which used the ThreadLocal hold  a reference to their own instance of StoreAppender which holds a reference back to the queue. It will be GCed only if the ThreadLocal is removed from all those threads, or when the threads die (which may be when the JVM process exits, depending on the application/framework that created the thread - specifically it is not a thread controlled by chronicle). This ThreadLocal is interesting since it is non-static, and there is a cyclic reference between the ThreadLocal and its value.

**Comment by peter-lawrey** [comment]
A circular reference should be able to be cleaned by the GC. We avoid
static ThreadLocal for the reasons you mention.
Once the queue and reference to be appender has been dropped do you see
that the Thread is holding a reference via the ThreadLocal even after a
full GC?
Java doesn't provide a supported way to clear ThreadLocal as you suggest
because it is supposed to clean these up  I know how to force it using
reflection if it is not.

Peter.

On 4 Nov 2016 00:23, "amichair" <notifications@github.com> wrote:

I think the queue will not be GCed, since the threads which used the
ThreadLocal hold a reference to their own instance of StoreAppender which
holds a reference back to the queue. It will be GCed only if the
ThreadLocal is removed from all those threads, or when the threads die
(which may be when the JVM process exits, depending on the
application/framework that created the thread - specifically it is not a
thread controlled by chronicle). This ThreadLocal is interesting since it
is non-static, and there is a cyclic reference between the ThreadLocal and
its value.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/OpenHFT/Chronicle-Queue/issues/305#issuecomment-258313395>,
or mute the thread

**Comment by amichair** [comment]
Yes, I see multiple copies of the entire application remain in memory when the application bundle is redeployed (in Karaf). A heap dump analysis showed the root of the whole caboodle is indeed the StoreAppender referenced by the ThreadLocal.

The GC does handle detached circular references properly, but I think in this case they are not detached - they are reachable from a GC root (iirc Threads are considered GC roots) and since they are reachable, they are not collected.

ThreadLocal.remove() removes the reference and allows it to be collected, but how exactly it is used depends on the application design. Another option often combined with ThreadLocals is using a thread pool, so that the lifecycle of the thread can be controlled by whomever sets the ThreadLocals, but again this depends on the application design. I'm not sure what the best strategy is for chronicle to manage these ThreadLocals, since they are not the textbook use case - perhaps digging into the implementation details and best practices will yield a good solution...

**Comment by peter-lawrey** [comment]
ThreadLocals in ThreadLocalMap use WeakReferences so once the ThreadLocal
has been discarded everything related to it is also discarded. If these
ThreadLOcal are being retained, either a Full GC hasn't been performed or
the reference is being held elsewhere.

From the source

    /**
     * ThreadLocalMap is a customized hash map suitable only for
     * maintaining thread local values. No operations are exported
     * outside of the ThreadLocal class. The class is package private to
     * allow declaration of fields in class Thread.  To help deal with
     * very large and long-lived usages, the hash table entries use
     * WeakReferences for keys. However, since reference queues are not
     * used, stale entries are guaranteed to be removed only when
     * the table starts running out of space.
     */
    static class ThreadLocalMap {

        /**
         * The entries in this hash map extend WeakReference, using
         * its main ref field as the key (which is always a
         * ThreadLocal object).  Note that null keys (i.e. entry.get()
         * == null) mean that the key is no longer referenced, so the
         * entry can be expunged from table.  Such entries are referred to
         * as "stale entries" in the code that follows.
         */
        static class Entry extends WeakReference<ThreadLocal<?>> {
            /** The value associated with this ThreadLocal. */
            Object value;

            Entry(ThreadLocal<?> k, Object v) {
                super(k);
                value = v;
            }
        }

**Comment by amichair** [comment]
It's a bit tricky... the weakly referenced instance is the key, which is the ThreadLocal instance itself, but the value is held using a regular reference (in the snippet above), and the value is the StoreAppender, which holds a reference to the queue, which holds a reference to the ThreadLocal instance... so if I understand this correctly, there is a chain (or rather cycle) of strong references to it from the Thread. It's a bit of an edge case in GC behavior if you draw out all the references including the weak reference and the GC root (thread), but from what I gather, it won't be collected. I may be wrong :-)

**Comment by peter-lawrey** [comment]
If the only way to reach a object is via a weak reference it should be
collected even if there is a reference back to the weak reference. The fact
the references are circular doesn't matter as it only marks which objects
are to be retained from the root, it doesn't do reference counting.

AFAIK If it's being retained, that means there is a soft or strong
reference (or the GC isn't being called)

**Comment by amichair** [comment]
> If the only way to reach a object is via a weak reference

This is not the case here - there are two ways to reach the object:
```
Thread --> ThreadLocalMap --> Entry --> (weak key) ThreadLocal
Thread --> ThreadLocalMap --> Entry --> (strong value) StoreAppender --> SingleChronicleQueue --> ThreadLocal
```
The (not too clear) WeakReference docs say a weak reference (top path) is cleared if there are no strong references to the same object. But since there is a strong reference to it (bottom path), the weak reference remains intact, and the ThreadLocal remains in the map, along with everything else. Is this correct?

**Comment by peter-lawrey** [comment]
I see now, the problem is not that the weak reference in the key but a lack
of a weak reference in the value. That's something I can fix.

**Comment by peter-lawrey** [comment]
This might solve the problem

```java
    public ExcerptAppender acquireAppender() {
        if (readOnly) {
            throw new IllegalStateException("Can't append to a read-only
chronicle");
        }
        WeakReference<ExcerptAppender> ref =
excerptAppenderThreadLocal.get();
        ExcerptAppender ret = null;
        if (ref != null) ret = ref.get();
        if (ret == null) {
            ret = newAppender();
            ref = new WeakReference<>(ret);
            excerptAppenderThreadLocal.set(ref);
        }

        return ret;
    }
```

**Comment by peter-lawrey** [comment]
I have added a template to core

public enum ThreadLocalHelper {
    ;

    public static <T> T getTL(ThreadLocal<WeakReference<T>> threadLocal,
Supplier<T> supplier) {
        WeakReference<T> ref = threadLocal.get();
        T ret = null;
        if (ref != null) ret = ref.get();
        if (ret == null) {
            ret = supplier.get();
            ref = new WeakReference<T>(ret);
            threadLocal.set(ref);
        }
        return ret;
    }

    public static <T, A> T getTL(ThreadLocal<WeakReference<T>> threadLocal,
A a, Function<A, T> function) {
        WeakReference<T> ref = threadLocal.get();
        T ret = null;
        if (ref != null) ret = ref.get();
        if (ret == null) {
            ret = function.apply(a);
            ref = new WeakReference<T>(ret);
            threadLocal.set(ref);
        }
        return ret;
    }
}

This way, not only the value but the Supplier is not retained (which is
important if you want to unload the code). There were objects in Bytes and
Wire which needed changing to use this template as well.

**Comment by peter-lawrey** [comment]
This a problem if;
- you don't cache/hold the value your self.
- it doesn't survive long enough to get into tenured space. After that only
a major/full collection will clean it up.

If you have a low latency application you will warm up the code and trigger
a full collection as you want a long time between minor GCs. Eg 24 hours or
a week ideally.

You are right that an application gc-ing often is made slightly worse by
this esp the reuse of buffers as they can be bigger.

**Comment by amichair** [comment]
Thanks for the quick responses!

Using a WeakReference for the value seems like a good solution. A few implementation thoughts:
- In getTL, in the case that the reference exists but its value is gone, there is no need to create a new WeakReference, only to set the value on the existing one (just a tiny optimization for less garbage.)
- I don't know what LongArrayValuesHolder is, but unless it references back to the ThreadLocal instance like in the Appender's case, there is no need for the fix there. The problem only occurs when there is a strong reference from the value back to the ThreadLocal that prevents it from being collected.
- This is a matter of taste, but perhaps instead of the getTL utility methods you can consider creating a WeakThreadLocal class that encapsulates the wrapping of the value with a weak reference, making it a simple replacement for ThreadLocal in the code that uses it.

**Comment by peter-lawrey** [comment]
AFAIK To support OSGI, you have to ensure an a thread local value not only
doesn't cause a memory leak but doesn't have a strong reference to a class
in a module you might want to unload/upgrade in the future.
Constructing the ThreadLocal this way should allow you to upgrade these
modules.

**Comment by amichair** [comment]
btw, if the value is always weakly referenced, won't it just be cleared and collected on the next GC run? Does the code that uses the value hold it (the strong reference obtained via get()) for long enough to make this worthwhile? Otherwise you'll have a lot of churn of value instances being created and destroyed every GC run, which may defeat the purpose of using a ThreadLocal in the first place...

**Comment by amichair** [comment]
I have another issue that may be related to this issue and/or fix... is it possible that cycled files aren't getting released? In a heap dump I see a ThreadLocal with SCQIndexing$LongArrayValuesHolder -> BinaryLongArrayReference -> MappedBytes -> MappedFile -> fd holding on to a file that was cycled out and (I think) should have been released but isn't. When should a file be ultimately released? Will the WeakReference fix above dispose of the reference but leave the file open (does a reference count need to be released?)

It's a bit hard to follow the code and there's not much documentation... could you please clarify the situation? If it's indeed an issue I can open a separate ticket for it, but I admit I'm not entirely sure how all these nested classes are managed.

**Comment by peter-lawrey** [comment]
It should need to be released before the file is closed.
I am not sure we have a test for this at the very least, so you can raise
another issue for it.

**Comment by peter-lawrey** [comment]
Closing this issue.
